packages/@react-spectrum/s2/src/Dialog.tsx

```diff
@@ -10,128 +10,127 @@
-    '@media (height < 400)': 'visible',
-    type: {
-      popover: 'visible'
-    }
-  },
-  font: 'body',
-  // TODO: adjust margin on mobile?
-  marginX: {
-    default: 32
-  }
-});
-
-const footer = style({
-  flexGrow: 1,
-  font: 'body'
-});
-
-const buttonGroup = style({
-  marginStart: 'auto',
-  maxWidth: 'full'
-});
-
-interface DialogContextValue extends Pick<AriaPopoverProps, 'placement' | 'shouldFlip' | 'containerPadding' | 'offset' | 'crossOffset'> {
-  /**
-   * The type of Dialog that should be rendered.
    */
-  type?: 'modal' | 'popover' | 'fullscreen' | 'fullscreenTakeover', // TODO: add tray back in
+  size?: 'S' | 'M' | 'L',
   /**
-   * Whether a modal type Dialog should be dismissable.
+   * Whether the Dialog is dismissable.
    */
   isDismissable?: boolean,
+  /** Whether pressing the escape key to close the dialog should be disabled. */
+  isKeyboardDismissDisabled?: boolean,
   /**
    * Whether a popover type Dialog's arrow should be hidden.
    */
   hideArrow?: boolean,
-  /**
-   * Whether pressing the escape key to close the dialog should be disabled.
-   */
-  isKeyboardDismissDisabled?: boolean
+  /** Whether the content of the Dialog should extend edge-to-edge, without any padding. */
+  isFullBleed?: boolean
 }
 
-export const DialogContext = createContext<DialogContextValue>({
-  type: 'modal',
-  isDismissable: false,
-  hideArrow: false,
-  shouldFlip: true,
-  isKeyboardDismissDisabled: false
-});
+const modalPadding = {
+  default: 24,
+  sm: 32
+} as const;
+
+const dialogStyle = style({
+  padding: {
+    type: {
+      modal: modalPadding,
+      popover: 8,
+      fullscreen: modalPadding,
+      fullscreenTakeover: modalPadding
+    },
+    isFullBleed: 0
+  },
+  boxSizing: 'border-box',
+  outlineStyle: 'none',
+  borderRadius: '[inherit]',
+  overflow: 'auto',
+  position: 'relative',
+  size: 'full',
+  maxSize: 'full'
+}, getAllowedOverrides({height: true}));
 
 function Dialog(props: DialogProps, ref: DOMRef) {
-  let ctx = useContext(DialogContext);
-  let isDismissable = ctx.isDismissable || props.isDismissable;
-  let isKeyboardDismissDisabled = ctx.isKeyboardDismissDisabled;
+  let {
+    type = 'modal',
+    mobileType = type === 'popover' ? 'modal' : type,
+    size,
+    isDismissable,
+    isKeyboardDismissDisabled,
+    isFullBleed
+  } = props;
   let domRef = useDOMRef(ref);
+  let stringFormatter = useLocalizedStringFormatter(intlMessages, '@react-spectrum/s2');
 
-  switch (ctx.type) {
+  let dialog = (
+    <RACDialog
+      {...props}
+      ref={domRef}
+      style={props.UNSAFE_style}
+      className={(props.UNSAFE_className || '') + dialogStyle({type, isFullBleed}, props.styles)}>
+      {composeRenderProps(props.children, (children) => (
+        // Reset OverlayTriggerStateContext so the buttons inside the dialog don't retain their hover state.
+        <OverlayTriggerStateContext.Provider value={null}>
```

Comment:
This came from the previous implementation, but I think it's due to S2 buttons in particular consuming `OverlayTriggerStateContext` to apply the hover state. If we put this in RAC it would prevent people from accessing the state inside all overlays which might not be expected.